File: age_group_agg_optimized.js
[
  {
    $match:
      // filter documents between male
      // and female where age group is
      // already defined
      {
        gender: {
          $in: ["M", "F"]
        },
        age_group: {
          $in: [
            "Under 30",
            "30–59",
            "60 and over"
          ]
        }
      }
  },
  {
    $group:
      // group by gender and age group
      {
        _id: {
          gender: "$gender",
          age_group: "$age_group"
        },
        count: {
          $sum: 1
        }
      }
  },
  {
    $sort:
      // sort by gender and age group
      // for cleaner output
      {
        "_id.gender": 1,
        "_id.age_group": 1
      }
  }
]

File: age_group_agg_optimized.py
from pymongo import MongoClient

# Requires the PyMongo package.
# https://api.mongodb.com/python/current

client = MongoClient('mongodb://localhost:27017/')
result = client['D597_optimization']['medical_optimization'].aggregate([
    {
        '$match': {
            'gender': {
                '$in': [
                    'M', 'F'
                ]
            }, 
            'age_group': {
                '$in': [
                    'Under 30', '30–59', '60 and over'
                ]
            }
        }
    }, {
        '$group': {
            '_id': {
                'gender': '$gender', 
                'age_group': '$age_group'
            }, 
            'count': {
                '$sum': 1
            }
        }
    }, {
        '$sort': {
            '_id.gender': 1, 
            '_id.age_group': 1
        }
    }
])